### PR TITLE
MGDSTRM-10879 adding control plane directed pause handling

### PR DESCRIPTION
**Pull request by shawkins**
Allows the control plane to pause syncing to the data plane via the pause annotation https://issues.redhat.com/browse/MGDSTRM-10879

The delete flag is still set to allow for cleanup to proceed.

This usage of the annotation will not intersect with sre usage of the pause annotation, but we can of course use an entirely different annotation key if anyone objects to this reuse.

### CODE REVIEW

File: sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java [hunk 1]
```diff
@@ -103,6 +103,25 @@ public void syncKafkaClusters() {
...
+                    continue;
+                }
+                if (existing != null) {
+                    if (!existing.getSpec().isDeleted()) {
+                        log.debugf("Passing just the delete flag to the ManagedKafka %s", Cache.metaNamespaceKeyFunc(remoteManagedKafka));
+                        kubeClient.resources(ManagedKafka.class)
+                                .inNamespace(existing.getMetadata().getNamespace())
+                                .withName(existing.getMetadata().getName())
+                                .patch(PatchContext.of(PatchType.JSON),
+                                        "[{\"op\": \"replace\", \"path\":\"/spec/deleted\", \"value\":true}]");
```
Comment (MikeEdgar) [inline sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: Why this style instead of using the POJO?

Comment (shawkins) [reply sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: Strategic merge is not supported.  There isn't a "good" direct way of producing the simplified json patch.  And json merge isn't great for partial documents, but may be fine here, I didn't check specifically.

**Review by machi1990 — APPROVED** [review]
lgtm, thanks @shawkins 

tested this PR from the control plane perspective.
let me know when the annotation key changes to something else.

**Review by MikeEdgar — APPROVED** [review]

File: sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java (hunk 1 above)
Comment (rareddy) [inline sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: little confused about the logic here, if the control plane pause is true, and if the local instance exists and is not deleted it sets "spec/delete" to true?

Comment (shawkins) [reply sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: You'll only do that if the remote is marked as deleted:

```
if remote paused {
  if remote not deleted {
    move to the next item to skip main reconciliation
  }
  if local exists {
    if local is not deleted {
      mark local deleted
    }
    move to the next item to skip main reconciliation
  }
} 
```

Comment (rareddy) [reply sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: Thanks! remote deleted and paused at the same time seems like a weird case imo.

Comment (machi1990) [reply sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: `deleted` is a status in the database. 
`pause-reconciliation` directly ties to whether some resources e.g TLS certificates where able to be fetched or not. These resources are fetched each time a ManagedKafka is reconciled with the Fleetshard sync irrespective of the status (wether deleted or not) of the Kafka.

Comment (rareddy) [reply sync/src/main/java/org/bf2/sync/ManagedKafkaSync.java:119]: ok. Thanks for explanation.

**Review by rareddy — APPROVED** [review]
lgtm

**Review by biswassri — APPROVED** [review]

**Comment by machi1990** [comment]
Thanks, will this be on the 0.35 release?

**Comment by shawkins** [comment]
@machi1990 yes it will be.

**Comment by machi1990** [comment]
Thanks

**Pull request merged**